07173 revert to netty

rbair23 opened the pull request:
Closes #7173

Removes Helidon and goes back to using base Netty. Changes to the mono service code were done by simple reverting the PRs that introduced Helidon. Changes to the modular code required a rewrite. The existing APIs used Helidon in the APIs, so modifications were made to be somewhat more generic. I moved Netty specific code into its own package, to try to make this a little more compartmentalized.

jsync-swirlds commented on hedera-node/hedera-app/build.gradle.kts:
```diff
@@ -52,13 +52,15 @@ dependencies {
         itestImplementation(gav("org.junit.jupiter.params"))
+        itestImplementation(gav("io.netty.transport.classes.epoll"))
+        itestImplementation(gav("io.netty.transport.epoll"))
+        itestImplementation("io.netty:netty-transport-native-epoll:4.1.94.Final")
```
This is different from what GRPC requires (4.1.87.Final).  We probably want to make sure the versions match to avoid more split-package issues.
Also, most versions are managed in platform now, so might be good to cross check that this matches that version (or better, manage the version there).

rbair23 replied:
Agreed, this is the same version in the BOM. I'm not sure why I had to call it out, but it wouldn't build otherwise. I could use @jjohannes's help here.

jjohannes replied:
This should not be needed.

As far as I can see, if I remove the following two lines, everything still works fine:

```
itestImplementation("io.netty:netty-transport-native-epoll:4.1.94.Final")
itestImplementation("io.netty:netty-transport-classes-epoll:4.1.94.Final")
```

What's see issue you have been seeing?

jsync-swirlds commented on hedera-platform/build.gradle.kts:
```diff
@@ -28,7 +28,7 @@ val helidonVersion = "3.2.1"
 val log4jVersion = "2.17.1"
 val mockitoVersion = "4.6.1"
-val nettyVersion = "4.1.87.Final"
+val nettyVersion = "4.1.94.Final"
```
GRPC 1.54.1 specifies Netty 4.1.87.Final.
We should probably keep those in sync (either use a newer GRPC, or the older netty).

rbair23 replied:
I'll update the version of grpc too

rbair23 commented:
The integration tests are failing because the node refuses to start within the docker container:

```2023-06-27 23:29:11.085 DEBUG 318  PlatformDependent0 - direct buffer constructor: unavailable
java.lang.UnsupportedOperationException: Reflective setAccessible(true) disabled
	at io.netty.util.internal.ReflectionUtil.trySetAccessible(ReflectionUtil.java:31) ~[netty-common-4.1.94.Final.jar:4.1.94.Final]
	at io.netty.util.internal.PlatformDependent0$5.run(PlatformDependent0.java:289) ~[netty-common-4.1.94.Final.jar:4.1.94.Final]
	at java.security.AccessController.doPrivileged(AccessController.java:318) ~[?:?]
	at io.netty.util.internal.PlatformDependent0.<clinit>(PlatformDependent0.java:282) ~[netty-common-4.1.94.Final.jar:4.1.94.Final]
	at io.netty.util.internal.PlatformDependent.isAndroid(PlatformDependent.java:333) ~[netty-common-4.1.94.Final.jar:4.1.94.Final]
	at io.netty.util.internal.PlatformDependent.<clinit>(PlatformDependent.java:88) ~[netty-common-4.1.94.Final.jar:4.1.94.Final]
	at io.netty.util.AsciiString.<init>(AsciiString.java:225) ~[netty-common-4.1.94.Final.jar:4.1.94.Final]
	at io.netty.util.AsciiString.<init>(AsciiString.java:212) ~[netty-common-4.1.94.Final.jar:4.1.94.Final]
	at io.netty.util.AsciiString.cached(AsciiString.java:1403) ~[netty-common-4.1.94.Final.jar:4.1.94.Final]
	at io.netty.util.AsciiString.<clinit>(AsciiString.java:48) ~[netty-common-4.1.94.Final.jar:4.1.94.Final]
	at io.grpc.netty.Utils.<clinit>(Utils.java:78) ~[grpc-netty-1.56.0-module.jar:1.56.0]
	at io.grpc.netty.NettyServerBuilder.<clinit>(NettyServerBuilder.java:84) ~[grpc-netty-1.56.0-module.jar:1.56.0]
	at com.hedera.node.app.service.mono.grpc.ConfigDrivenNettyFactory.builderFor(ConfigDrivenNettyFactory.java:67) ~[app-service-mono-0.40.0-SNAPSHOT.jar:?]
	at com.hedera.node.app.service.mono.grpc.NettyGrpcServerManager.startOneNettyServer(NettyGrpcServerManager.java:84) ~[app-service-mono-0.40.0-SNAPSHOT.jar:?]
	at com.hedera.node.app.service.mono.grpc.NettyGrpcServerManager.start(NettyGrpcServerManager.java:69) ~[app-service-mono-0.40.0-SNAPSHOT.jar:?]
	at com.hedera.node.app.service.mono.grpc.GrpcStarter.startIfAppropriate(GrpcStarter.java:81) ~[app-service-mono-0.40.0-SNAPSHOT.jar:?]
	at com.hedera.node.app.MonoServicesMain.startNettyIfAppropriate(MonoServicesMain.java:119) ~[?:?]
	at com.hedera.node.app.MonoServicesMain.doStagedInit(MonoServicesMain.java:110) ~[?:?]
	at com.hedera.node.app.MonoServicesMain.initApp(MonoServicesMain.java:90) ~[?:?]
	at com.hedera.node.app.MonoServicesMain.init(MonoServicesMain.java:68) ~[?:?]
	at com.hedera.node.app.ServicesMain.init(ServicesMain.java:69) ~[?:?]
	at com.swirlds.platform.Browser.createLocalPlatforms(Browser.java:697) ~[swirlds-platform-core-0.39.0-alpha.3.jar:?]
	at com.swirlds.platform.Browser.startPlatforms(Browser.java:830) ~[swirlds-platform-core-0.39.0-alpha.3.jar:?]
	at com.swirlds.platform.Browser.<init>(Browser.java:312) ~[swirlds-platform-core-0.39.0-alpha.3.jar:?]
	at com.swirlds.platform.Browser.launch(Browser.java:545) ~[swirlds-platform-core-0.39.0-alpha.3.jar:?]
	at com.swirlds.platform.Browser.parseCommandLineArgsAndLaunch(Browser.java:496) ~[swirlds-platform-core-0.39.0-alpha.3.jar:?]
	at com.swirlds.platform.Browser.main(Browser.java:854) ~[swirlds-platform-core-0.39.0-alpha.3.jar:?]
2023-06-27 23:29:11.088 DEBUG 386  PlatformDependent0 - java.nio.Bits.unaligned: available, true
2023-06-27 23:29:11.089 DEBUG 450  PlatformDependent0 - jdk.internal.misc.Unsafe.allocateUninitializedArray(int): unavailable
java.lang.IllegalAccessException: class io.netty.util.internal.PlatformDependent0$7 cannot access class jdk.internal.misc.Unsafe (in module java.base) because module java.base does not export jdk.internal.misc to unnamed module @5ac154b6
	at jdk.internal.reflect.Reflection.newIllegalAccessException(Reflection.java:392) ~[?:?]
	at java.lang.reflect.AccessibleObject.checkAccess(AccessibleObject.java:674) ~[?:?]
	at java.lang.reflect.Method.invoke(Method.java:560) ~[?:?]
	at io.netty.util.internal.PlatformDependent0$7.run(PlatformDependent0.java:411) ~[netty-common-4.1.94.Final.jar:4.1.94.Final]
	at java.security.AccessController.doPrivileged(AccessController.java:318) ~[?:?]
	at io.netty.util.internal.PlatformDependent0.<clinit>(PlatformDependent0.java:402) ~[netty-common-4.1.94.Final.jar:4.1.94.Final]
	at io.netty.util.internal.PlatformDependent.isAndroid(PlatformDependent.java:333) ~[netty-common-4.1.94.Final.jar:4.1.94.Final]
	at io.netty.util.internal.PlatformDependent.<clinit>(PlatformDependent.java:88) ~[netty-common-4.1.94.Final.jar:4.1.94.Final]
	at io.netty.util.AsciiString.<init>(AsciiString.java:225) ~[netty-common-4.1.94.Final.jar:4.1.94.Final]
	at io.netty.util.AsciiString.<init>(AsciiString.java:212) ~[netty-common-4.1.94.Final.jar:4.1.94.Final]
	at io.netty.util.AsciiString.cached(AsciiString.java:1403) ~[netty-common-4.1.94.Final.jar:4.1.94.Final]
	at io.netty.util.AsciiString.<clinit>(AsciiString.java:48) ~[netty-common-4.1.94.Final.jar:4.1.94.Final]
	at io.grpc.netty.Utils.<clinit>(Utils.java:78) ~[grpc-netty-1.56.0-module.jar:1.56.0]
	at io.grpc.netty.NettyServerBuilder.<clinit>(NettyServerBuilder.java:84) ~[grpc-netty-1.56.0-module.jar:1.56.0]
	at com.hedera.node.app.service.mono.grpc.ConfigDrivenNettyFactory.builderFor(ConfigDrivenNettyFactory.java:67) ~[app-service-mono-0.40.0-SNAPSHOT.jar:?]
	at com.hedera.node.app.service.mono.grpc.NettyGrpcServerManager.startOneNettyServer(NettyGrpcServerManager.java:84) ~[app-service-mono-0.40.0-SNAPSHOT.jar:?]
	at com.hedera.node.app.service.mono.grpc.NettyGrpcServerManager.start(NettyGrpcServerManager.java:69) ~[app-service-mono-0.40.0-SNAPSHOT.jar:?]
	at com.hedera.node.app.service.mono.grpc.GrpcStarter.startIfAppropriate(GrpcStarter.java:81) ~[app-service-mono-0.40.0-SNAPSHOT.jar:?]
	at com.hedera.node.app.MonoServicesMain.startNettyIfAppropriate(MonoServicesMain.java:119) ~[?:?]
	at com.hedera.node.app.MonoServicesMain.doStagedInit(MonoServicesMain.java:110) ~[?:?]
	at com.hedera.node.app.MonoServicesMain.initApp(MonoServicesMain.java:90) ~[?:?]
	at com.hedera.node.app.MonoServicesMain.init(MonoServicesMain.java:68) ~[?:?]
	at com.hedera.node.app.ServicesMain.init(ServicesMain.java:69) ~[?:?]
	at com.swirlds.platform.Browser.createLocalPlatforms(Browser.java:697) ~[swirlds-platform-core-0.39.0-alpha.3.jar:?]
	at com.swirlds.platform.Browser.startPlatforms(Browser.java:830) ~[swirlds-platform-core-0.39.0-alpha.3.jar:?]
	at com.swirlds.platform.Browser.<init>(Browser.java:312) ~[swirlds-platform-core-0.39.0-alpha.3.jar:?]
	at com.swirlds.platform.Browser.launch(Browser.java:545) ~[swirlds-platform-core-0.39.0-alpha.3.jar:?]
	at com.swirlds.platform.Browser.parseCommandLineArgsAndLaunch(Browser.java:496) ~[swirlds-platform-core-0.39.0-alpha.3.jar:?]
	at com.swirlds.platform.Browser.main(Browser.java:854) ~[swirlds-platform-core-0.39.0-alpha.3.jar:?]```

rbair23 commented:
Found and fixed the problem with the docker container -- the epoll native libraries were not included in the build dependencies, and thus were not in the image. In addition, some flags were set to work with the modular libraries when running the java app. I also changed the docker build process so it is more streamlined and didn't rebuild everything when building the docker image, but reuses the built libraries instead. This makes it MUCH faster to build a docker image.

jsync-swirlds commented on hedera-node/test-clients/src/main/java/com/hedera/services/bdd/spec/infrastructure/HapiApiClients.java:
```diff
@@ -118,154 +111,75 @@ private ManagedChannel createNettyChannel(boolean useTls, final String host, fin
 
-    private String workflowStubId(AccountID nodeId, boolean useTls) {
-        return useTls ? workflowTlsStubIds.get(nodeId) : workflowStubIds.get(nodeId);
+    private static ChannelStubs randomStubsFromPool(@NonNull final List<ChannelStubs> stubs) {
```
Just a thought here.  Random selection still leaves a roughly 1:50 chance we double up on a channel.  Would it be better to round-robin these so the probability of any given call being on the same channel as another is minimized?

nathanklick commented on hedera-node/docker/start-services.sh:
```diff
@@ -68,4 +68,7 @@ if [[ "${CI_WAIT_FOR_PEERS}" = true && -n "${NODE_ID}" ]]; then
 fi
 
 java -cp 'data/lib/*' -Dflag=1 -Dfile.encoding='utf-8' \
+  --add-opens java.base/jdk.internal.misc=ALL-UNNAMED \
+  --add-opens java.base/java.nio=ALL-UNNAMED \
+  -Dio.netty.tryReflectionSetAccessible=true \
```
Do we need to ensure the production containers use these options to bypass the netty JPMS issues? If so, then a ticket should be sent to DevOps for the appropriate NMT changes to make these the default.

jsync-swirlds replied:
We will.  Where do I enter that ticket?

tinker-michaelj reviewed (approved):
LGTM!
